File: backend/internal/models/store.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"

	"github.com/google/uuid"
)

type Store struct {
	ID            uuid.UUID        `json:"id" db:"id"`
	Name          string           `json:"name" db:"name"`
	Address       string           `json:"address" db:"address"`
	Latitude      float64          `json:"latitude" db:"latitude"`
	Longitude     float64          `json:"longitude" db:"longitude"`
	Categories    StringArray      `json:"categories" db:"categories"`
	BusinessHours BusinessHoursData `json:"business_hours" db:"business_hours"`
	ParkingInfo   string           `json:"parking_info" db:"parking_info"`
	WebsiteURL    string           `json:"website_url" db:"website_url"`
	GoogleMapURL  string           `json:"google_map_url" db:"google_map_url"`
	SnsUrls       StringArray      `json:"sns_urls" db:"sns_urls"`
	Tags          StringArray      `json:"tags" db:"tags"`
	Photos        StringArray      `json:"photos" db:"photos"`
	CreatedBy     uuid.UUID        `json:"created_by" db:"created_by"`
	CreatedAt     time.Time        `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time        `json:"updated_at" db:"updated_at"`
}

// BusinessHoursData represents detailed business hours structure
type BusinessHoursData struct {
	Monday    DaySchedule `json:"monday"`
	Tuesday   DaySchedule `json:"tuesday"`
	Wednesday DaySchedule `json:"wednesday"`
	Thursday  DaySchedule `json:"thursday"`
	Friday    DaySchedule `json:"friday"`
	Saturday  DaySchedule `json:"saturday"`
	Sunday    DaySchedule `json:"sunday"`
}

// DaySchedule represents a day's schedule with up to 3 time slots
type DaySchedule struct {
	IsClosed   bool       `json:"is_closed"`
	TimeSlots  []TimeSlot `json:"time_slots"` // Maximum 3 slots
}

// TimeSlot represents one time period in a day
type TimeSlot struct {
	OpenTime      string `json:"open_time"`       // HH:MM format
	CloseTime     string `json:"close_time"`      // HH:MM format
	LastOrderTime string `json:"last_order_time"` // HH:MM format
}

// Value implements driver.Valuer for BusinessHoursData
func (b BusinessHoursData) Value() (driver.Value, error) {
	return json.Marshal(b)
}

// Scan implements sql.Scanner for BusinessHoursData
func (b *BusinessHoursData) Scan(value interface{}) error {
	if value == nil {
		*b = GetDefaultBusinessHours()
		return nil
	}

	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("cannot scan business_hours: unsupported type")
	}

	return json.Unmarshal(bytes, b)
}

// GetDefaultBusinessHours returns default business hours structure
func GetDefaultBusinessHours() BusinessHoursData {
	return BusinessHoursData{
		Monday:    DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Tuesday:   DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Wednesday: DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Thursday:  DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Friday:    DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Saturday:  DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
		Sunday:    DaySchedule{IsClosed: false, TimeSlots: []TimeSlot{}},
	}
}



type StringArray []string

func (a StringArray) Value() (driver.Value, error) {
	return json.Marshal(a)
}

func (a *StringArray) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}

	return json.Unmarshal(bytes, a)
}

File: frontend/src/types/store.ts
export interface Store {
  id: string
  name: string
  address: string
  latitude: number
  longitude: number
  categories: string[]
  business_hours: BusinessHoursData
  parking_info: string
  website_url: string
  google_map_url: string
  sns_urls: string[]
  tags: string[]
  photos: string[]
  created_by: string
  created_at: string
  updated_at: string
}

export interface Review {
  id: string
  store_id: string
  user_id: string
  rating: number
  comment: string | null
  photos: string[]
  visit_date: string | null
  is_visited: boolean
  payment_amount: number | null
  food_notes: string | null
  created_at: string
  updated_at: string
  user?: {
    id: string
    username: string
  }
}

export interface User {
  id: string
  username: string
  email: string
  role: 'admin' | 'editor' | 'viewer'
  created_at: string
  updated_at: string
}

export interface FilterState {
  name: string
  categories: string[]
  categoriesOperator: 'AND' | 'OR'
  tags: string[]
  tagsOperator: 'AND' | 'OR'
  businessDay: string
  businessTime: string
}

export interface APIResponse<T = any> {
  success: boolean
  data?: T
  error?: {
    code: string
    message: string
    details?: string
  }
  meta?: {
    total?: number
    page?: number
    limit?: number
    offset?: number
  }
}

export interface LocationCoordinates {
  latitude: number
  longitude: number
}

export interface TimeSlot {
  open_time: string
  close_time: string
  last_order_time: string
}

export interface DaySchedule {
  is_closed: boolean
  time_slots: TimeSlot[]
}

export interface BusinessHoursData {
  monday: DaySchedule
  tuesday: DaySchedule
  wednesday: DaySchedule
  thursday: DaySchedule
  friday: DaySchedule
  saturday: DaySchedule
  sunday: DaySchedule
}

export interface BusinessHoursInput {
  day: string
  time: string
}

export interface ValidationError {
  field: string
  message: string
}

export interface StoreFormData {
  name: string
  address: string
  latitude: number
  longitude: number
  categories: string[]
  business_hours: BusinessHoursData
  parking_info: string
  website_url: string
  google_map_url: string
  sns_urls: string[]
  tags: string[]
  photos: string[]
}

export interface ReviewFormData {
  store_id: string
  rating: number
  comment: string
  photos: string[]
  visit_date: Date | null
  is_visited: boolean
  payment_amount: number | undefined
  food_notes: string
}

export interface AuthContextType {
  user: User | null
  login: (email: string, password: string) => Promise<void>
  logout: () => void
  loading: boolean
  hasRole?: (role: string) => boolean
  isAdmin?: () => boolean
  isEditor?: () => boolean
  canEdit?: (createdBy: string) => boolean
}

export const USER_ROLES = {
  ADMIN: 'admin' as const,
  EDITOR: 'editor' as const,
  VIEWER: 'viewer' as const,
} as const

export const DEFAULT_FILTER_STATE: FilterState = {
  name: '',
  categories: [],
  categoriesOperator: 'OR',
  tags: [],
  tagsOperator: 'AND',
  businessDay: '',
  businessTime: '',
} as const